README.md
# Sym — Light-weight Symmetric Encryption for Humans

[![Gem Version](https://badge.fury.io/rb/sym.svg)](https://badge.fury.io/rb/sym)
[![Sym Downloads](http://ruby-gem-downloads-badge.herokuapp.com/sym?extension=svg)](https://rubygems.org/gems/sym)
[![Build Status](https://travis-ci.org/kigster/sym.svg?branch=master)](https://travis-ci.org/kigster/sym)
[![CircleCI](https://circleci.com/gh/kigster/sym.svg?style=svg)](https://circleci.com/gh/kigster/sym)

[![Maintainability](https://api.codeclimate.com/v1/badges/4f1d1614ccaa61c974dd/maintainability)](https://codeclimate.com/github/kigster/sym/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/4f1d1614ccaa61c974dd/test_coverage)](https://codeclimate.com/github/kigster/sym/test_coverage)
[![Issue Count](https://codeclimate.com/github/kigster/sym/badges/issue_count.svg)](https://codeclimate.com/github/kigster/sym)
[![Gitter](https://img.shields.io/gitter/room/gitterHQ/gitter.svg)](https://gitter.im/kigster/sym)


---

Please checkout the post "**[Dead Simple Encryption with Sym](http://kig.re/2017/03/10/dead-simple-encryption-with-sym.html)**" that announces this library, and provides further in-depth discussion. Your donation of absolutely any amount is very much appreciated.

[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=FSFYYNEQ8RKWU)
 
## Sym — Symmetric Encryption that's easy on Humans

<div style="padding 40px; margin: 40px; font-size: 13pt;">

<ul>
<li><strong>sym</strong> is an open source command line utility and Ruby API which makes it very <em>easy to add reliable encryption and decryption</em> of sensitive data to an application or a project written in any language.</li><br />

<li>Unlike many existing encryption tools, <strong>sym</strong> focuses on narrowing the gap between convenience and security, by offering enhanced usability and a streamlined ruby API and a CLI. The primary goal of the library is to make encryption very easy and transparent. </li><br />

<li><strong>sym</strong> uses the <em><a href="https://en.wikipedia.org/wiki/Symmetric-key_algorithm">Symmetric Encryption</a></em> algorithm. This means that the same key is used to encrypt and decrypt data. In addition to the key, the encryption uses a randomized IV vector, which is automatically generated  per each encryption and serialized with the data. Result of encryption is zlib-compressed, and base64 encoded, to be suitable for storage as string. The generated keys are also base64-encoded for convenience.</li><br />

<li>Finally, the library offers encryption using any regular password, and in particular supports password-protected encryption keys. Automatic key detection algorithm attempts to resolve a provided key as a filename, an environment variable name, an OS-X Keychain password entry name, a key itself, or a default key file.</li><br />

<li><strong>NOTE:</strong> Sym uses Ruby's <code>Marshall.dump</code> to serialize it's data, and therefore it is not currently possible or easy to deserialize the data in languages other than Ruby.</li>
</ul>
</div>

### Quick Demo of the CLI in Action

[![asciicast](design/ascii-cinema.png)](https://asciinema.org/a/106737)


#### Help Screens, Examples and Symit Bash Wrapper

This may be a good time to take a look at the full help message for the `sym` tool, shown naturally with a `-h` or `--help` option. Examples can be shown with `-E/--examples` flag. 

Additionally, Sym comes with a helpful BASH wrapper `symit`. 

**Help screens for `sym` and `symit` are shown in full on another page — [Sym Help Screens and Symit](SYM-CLI.md). Please refer to it for complete help screens and the examples.**


## Supported Ruby Versions

[![Build Status](https://travis-ci.org/kigster/sym.svg?branch=master)](https://travis-ci.org/kigster/sym)

Sym currently builds and runs on the following ruby versions:

 * 2.2.9
 * 2.3.6
 * 2.4.3
 * 2.5.1
 * jruby-9.1.7.0 

### Motivation

The main goal when writing this tool was to streamline and simplify handling of sensitive data in a  trasparent and easy to use way without sacrificing security.

Most common use-cases include:

 * __Encrypting/decrypting of application secrets files__, so that the encrypted secrets can be safely checked into the git repository and distributed, and yet without much of the added headache that this often requires
 * __Secure message transfer between any number of receipients__
 * __General purpose encryption/decryption with a 256-bit encryption key__, optionally itself re-encrypted with a password.
 * __General purpose encryption/decryption with an arbitrary password__.

__Sym__ is a layer built on top of the [`OpenSSL`](https://www.openssl.org/) library, and, hopefully, makes encryption more accessible to every-day developers, QA, and dev-ops folks, engaged in deploying applications.

### What's Included

This gem includes two primary components:

 1. **[Rich command line interface CLI](#cli)** with many features to streamline encryption/decryption, and to be integrated into the deployment flow.<br /><br />
 2.  Ruby APIs:
     * **[Key Generation, Encryption & Decryption API](#rubyapi)**
       - is activated by including `Sym` module in a class, it adds easy to use `encr`/`decr` methods.     
     * **[Application API to shadow the CLI usage](#rubyapi-app)**
       - You can instantiate `Sym::Application` class with a hash representing CLI arguments, and then call it's `#execute` method to mimic CLI execution.
     * **[Sym::MagicFile API](#magic-file)**
       - This is a convenience class allowing you to encrypt/decrypt files in your ruby code with just couple of lines of code.
     * **[Sym::Configuration](#rubyapi-config)** 
       - Use this class to override the default cipher, and configure other parameters such as compression, password caching, and more.

### Massive Time Savers

**Sym** tries very hard to get out of your way, to make it *feel* as if your encrypted files are as easy to work with as the unencrypted files. It accomplishes this transparency with the following features:
 
  * By using **Mac OS-X Keychain**, `sym` offers a simple yet secure way of storing the key on a local machine, much more secure then storing it on a file system.
  * By using a **password cache** (`-c`) via an in-memory provider such as `memcached`, `sym` invocations take advantage of password cache, and only ask for a password once per a configurable time period.
  * By using **`SYM_ARGS` environment variable** you can save common flags and they will be applied whenever `-A` flag is activated.
  * By reading a key from the default key source file `~/.sym.key` which requires no flags at all.
  * By utilizing the **`--negate` option to quickly encrypt a regular file**, or decrypt an encrypted file with extension `.enc`.
  * By using the **`-t file` (edit) mode**, that opens an encrypted file in your `$EDITOR`, and replaces the encrypted version upon save & exit.

As you can see, we really tried to build a tool that provides good security for application secrets, including password-based encryption, but does not annoyingly ask for password every time. With `--edit` option, and `--negate` options you can treat encrypted files like regular files. 

> Encrypting application secrets had never been easier! –– Socrates [LOL, -ed.]

## Using Sym

#### Installation

If you plan on using the library in your Ruby project with Bundler managing its dependencies, just include the following line in your `Gemfile`:

    gem 'sym'

And then run `bundle`.

Or install it into the global namespace with `gem install` command:

    $ gem install sym
    $ sym -h
    $ sym -E # see examples

__BASH Completion__

Optionally, after gem installation, you can also install bash-completion of gem's command line options, but running the following command (and feel free to use any of the "dot" files you prefer):

    sym -B ~/.bashrc

Should you choose to install it (this part is optional), you will be able to use "tab-tab" after typing `sym`, and you'll be able to choose from all of the supported flags.

#### Typical Use-Case Scenario

  1. You generate a new encryption key, that will be used to both encrypt and decrypt the data. The key is 256 bits, or 32 bytes, or 45 bytes when base64-encoded, and can be generated with `sym -g`. The key must be saved somewhere for later retrieval. The key should not be easily accessible to an attacker. Note, that while generating the key, you can:
     * optionally password protect the key with `sym -gp`
     * save the key into a file with `sym -gpo key-file` 
     * save it into the OS-X Keychain, with `sym -gpx keychain-name`
     * cache the password, with `sym -gpcx keychain-name`
     * Normally, `sym` will print the resulting key to STDOUT
     * You can prevent the key from being printed to STDOUT with `-q/--quiet`. 
  2. Next, let's assume you have a file or a string that you want to encrypt. We call this *data*.
  3. In order to encrypt the __data__, we must supply an encryption key. Flag `-k` automatically retrieves the key, by trying to read it in several distinct ways, such as:
     * a file with a pathname specified by the argument (eg, `-k ~/.key`)
     * or environment variable (eg `-k ENC_KEY`)
     * or OS-X Keychain entry
     * verbatum string argument (not recommended)
     * alternatively, you can paste the key interactively with `-i` or save the default key in `~/.sym.key` file.
  4. Finally, we are ready to encrypt. The data to be encrypted can be read from a file with `-f filename`, or it can be read from STDIN, or a passed on the command line with `-s string`. For example, `sym -e -k ~/.key -f /etc/passwd` will encrypt the file and print the encrypted contents to STDOUT.
  4. Instead of printing to STDOUT, the output can be saved to a file with `-o <file>` or a simple redirect or a pipe.
  5. Encrypted file can later be decrypted with `sym -d ...` assuming the same key it was encrypted with.
  6. Encrypted file with extension `.enc` can be automatically decrypted with `-n/--negate file` option; if the file does not end with `.enc`, it is encrypted and `.enc` extension added to the resulting file.
  7. With `-t/--edit file` flag you can edit an encrypted file in VIM (or `$EDITOR`) any encrypted file and edit it. Once you save it, the file gets re-encrypted and replaces the previous version. A backup can be created with `-b` option. See the section on [inline editing](#inline)

A sample session that uses Mac OS-X Keychain to store the password-protected key.

```bash
# Gen a new key, password-encrypt it, cache the password, save
# result in the key chain entry 'my-new-key' (but don't print it '-q')
❯ sym -gpqcx my-new-key
New Password     :  •••••••••
Confirm Password :  •••••••••

❯ sym -eck my-new-key -s 'My secret data' -o secret.enc
Password: •••••••••

❯ cat secret.enc
BAhTOh1TeW06OkRhdGE6OldyYXBFefDFFD.....

❯ sym -dck my-new-key -f secret.enc
My secret data

# Now, let's save our keychain key in the default key file:
❯ sym -ck my-new-key -o ~/.sym.key
# Now we can decrypt/encrypt with this key at will
❯ sym -n secret.enc
# created a decrypted file `secret` 

# Lets now save common flags in the SYM_ARGS bash variable:
❯ export SYM_ARGS="-ck my-new-key"
❯ sym -df secret.enc 
My secret data
```

Note that password caching is off by default, but is enabled with `-c` flag. In the example above, the decryption step fetched the password from the cache, and so the user was not required to re-enter the password.

<a name="inline"></a>

#### Inline Editing of Encrypted Files

The `sym` CLI tool supports one particularly interesting mode, that streamlines handling of encrypted files. The mode is called __edit mode__, and is activated with the `-t` flag. 

Instead of decrypting data anytime you need to change it into a new file and then manually re-encrypting the result, you can use the shortcut flag `-t` (for "edi**t**"), which decrypts your data into a temporary file, automatically opening it with an `$EDITOR`. 

    sym -t config/application/secrets.yml.enc -k ~/.key
    
> This is one of those time-saving features that can make a difference in making encryption feel easy and transparent.

> NOTE: this mode does not seem to work with GUI editors such as Atom or TextMate. Since `sym` waits for the editor process to complete, GUI editors "complete" immediately upon starting a windowed application. 

In this mode several flags are of importance:

    -b (--backup)   – will create a backup of the original file
    -v (--verbose) - will show additional info about file sizes

Here is a full command that opens a file specified by `-f | --file`, using the key specified in `-k | --keyfile`, in the editor defined by the `$EDITOR` environment variable (or if not set – defaults to `/bin/vi`)".

Example: here we edit an encrypted file in `vim`, while using interactive mode to paste the key (`-i | --interactive`), and then creating a backup file (`-b | --backup`) upon save:

    sym -ibt data.enc
    # => Private Key: ••••••••••••••••••••••••••••••••••••••••••••
    #
    # => Diff:
    # 3c3
    # # (c) 2015 Konstantin Gredeskoul.  All rights reserved.
    # ---
    # # (c) 2016 Konstantin Gredeskoul.  All rights reserved.

Note the `diff` shown after save.

<a name="rubyapi"></a>

## Ruby API

### Including `Sym` module

Low-level encryption routines can be imported by including `Sym` module into your class or a module. Such class will be decorated with new class methods `#private_key` and `#create_private_key`, as well as instance methods `#encr`, and `#decr`.

#### Class Method `#create_private_key()` 

This method will generate a new key each time it's called.

#### Class Method `#private_key(value = nil)` 

This method will either assign an existing key (if a value is passed) or generate and save a new key in the class instance variable. Therefore each class including `Sym` will (by default) use a unique key (unless the key is passed in as an argument).

The following example illustrates this point:

```ruby
require 'sym'

class TestClass
  include Sym
end

@key = TestClass.create_private_key
@key.eql?(TestClass.private_key)  # => false
# A new key was created and saved in #private_key accessor.

class SomeClass
  include Sym
  private_key TestClass.private_key
end

@key.eql?(SomeClass.private_key)  # => true (it was assigned)
```

#### Encrypting and Decrypting

So how would we use this library from another Ruby project to encrypt and decrypt values?

After including the `Sym` module, two instance methods are added: 

* `#encr(value, private_key)` and 
* `#decr(value, private_key)`.

Therefore you could write something like this below, protecting a sensitive string using a class-level secret.

```ruby
require 'sym'
class TestClass
  include Sym
  private_key ENV['SECRET']

  def sensitive_value=(value)
    @sensitive_value = encr(value, self.class.private_key)
  end
  def sensitive_value
    decr(@sensitive_value, self.class.private_key)
  end
end
```

#### Encrypting the Key Itself

You can encrypt the private key using a custom password. This is highly recommended, because without the password the key is the only piece that stands between an attacker and decrypting your sensitive data. 

For this purpose, two more instance methods exist:

 * `#encr_password(data, password, iv = nil)`
 * `#decr_password(encrypted_data, password, iv = nil)`

They can be used independently of `encr` and `decr` to encrypt/decrypt any data with a password.

<a name="magic-file"></a>

### Using `Sym::MagicFile` API for Reading/Writing Encrypted/Decrypted data

This is probably the easiest way to leverage Sym-encrypted files in your application — by loading them into memory with `Sym::MagicFile`. This class provides a very simple API while supporting all of the convenience features of the rich application API (see below).

You instantiate `Sym::MagicFile` with just two parameters: a `pathname` to a file (encrypted
or not), and the `key` identifier. The identifier can either be a filename, or
OS-X Keychain entry, or environment variable name, etc — basically it is resolve
like any other `-k <value>` CLI flag.

The following methods are available:

 * `#encrypt` — returns an encrypted string representing the encrypted contents ofa file specified by the pathname.
 * `#decrypt` — returns a decrypted string representing the decrypted contents of a file specified by  the pathname.
 * `#encrypt_to(filename)` — encrypts the contents of a file specified by the pathname, and writes the result to a `filename`. 
 * `#decrypt_to(filename)` — decrypts the contents of a file specified by the pathname, and writes the result to a `filename`. 

#### Example: Using `Sym::MagicFile` with the `RailsConfig` (or `Settings`) gem

In this example, we assume that the environment variable `$PRIVATE_KEY` contain
the key to be used in decryption. 

```ruby
require 'sym/magic_file'
require 'yaml'
secrets = Sym::MagicFile.new('/usr/local/etc/secrets.yml.enc', 'PRIVATE_KEY')
hash = YAML.load(secrets.decrypt)
```

Let's say that you are using [RailsConfig](https://github.com/railsconfig/config) gem for managing your Rails application setings. Since the gem allows appending settings from a hash, you can simply do the following in your `settings_initializer.rb`, and after all of the unencrypted settings are loaded:

```ruby
require 'config'
require 'sym/magic_file'
require 'yaml'
Settings.add_source!(
    YAML.load(
        Sym::MagicFile.new(
            '/usr/local/etc/secrets.yml.enc', 
            'PRIVATE_KEY'
        ).decrypt)
    )
Settings.reload!
```

<a name="rubyapi-app"></a>

### Using `Sym::Application` API

Since the command line interface offers much more than just encryption/decryption of data with a key, majority of these features are available through `Sym::Application` instance.

The class is instantiated with a hash that would be otherwise generated by parsing CLI arguments, typical `options`. For example, to generate the key, pass `generate: true` — essentially any flag in it's long form can be converted into a hash member.

Here is an example:

```ruby
require 'sym/application'

key  = Sym::Application.new(generate: true).execute
# => '75ngenJpB6zL47/8Wo7Ne6JN1pnOsqNEcIqblItpfg4='
```

### Ruby API Conclusion

Using `Sym`'s rich ruby API you can perform both low-level encryption/decryption, as well as high-level management of encrypted files. By using `Sym::MagicFile` and/or `Sym::Application` classes you can access the entire set of functionality expressed vi the CLI, described in details below.

<a name="cli"></a>
## Using `sym` with the Command Line

### Encryption Keys

The private key is the cornerstone of the symmetric encryption. Using `sym`, the key can be:

 * generated and printed to STDOUT, or saved to Mac OS-X KeyChain or a file
 * fetched from the Keychain in subsequent operations
 * password-protected during generation (or import) with the `-p` flag.
 * password can be cached using a locally running `memcached`, assuming the `-c` flag is provided.
 * must be kept very well protected and secure from attackers.

The __unencrypted private__ key will be in the form of a base64-encoded string, 45 characters long.

__Encrypted (with password) private key__ will be considerably longer, perhaps 200-300 characters long.

#### Generating the Key — Examples

```bash
# Let's generate a new key, and copy it to the clipboard (using `pbcopy` command on Mac OS-X):
$ sym -g | pbcopy

# Or save a new key into a bash variable
$ KEY=$(sym -g)

# Or save it to a file:
$ sym -go ~/.key

# Or create a password-protected key (`-p`), and save it to a file (`-o`), 
# cache the password (`-c`), and don't print the new key to STDOUT (`-q` for quiet)
$ sym -gpcqo ~/.secret
New Password:     ••••••••••
Confirm Password: ••••••••••
$ 
```

#### Resolving the `-k` Argument
    
You can use the generated private key by passing an argument to the `-k` flag. 

**Sym** attempts to automatically resolve the key source by trying each of the following options, and then moving on to the next until the key is found, or error is shown:

 1. the `-k value` flag, where the *value* is one of:
    * a file path, eg (`-k ~/.key`)
    * an environment variable name (`-k MY_KEY`)
    * an actual base64-encoded key (not recommended for security reasons)
    * a keychain name (`-k keychain-entry-name`)
 2. pasting or typing the key with the `-i` (interactive) flag
 3. if exists, a default key file, located in your home folder: `~/.sym.key` is used only when no other key-specifying flags were passed in.

#### Encryption and Decryption

<a name="inline"></a>

#### Inline Editing

The `sym` CLI tool supports one particularly interesting mode, that streamlines handling of encrypted files. The mode is called __edit mode__, and is activated with the `-t file` flag. 

In this mode `sym` will automaticaly decrypt the encrypted file into a temporary file, and then open it in `$EDITOR`. Once you quit the editor, `sym` will automatically diff the new and old content, and if it is different, `sym` will re-encrypt the new contents and overwrite the original file. You can create an optional backup by adding `-b` flag.

> NOTE: this mode does not seem to work with GUI editors such as Atom or TextMate. Since `sym` waits for the editor process to complete, GUI editors "complete" immediately upon starting a windowed application. 
In this mode several flags are of importance:

    -b (--backup)   – will create a backup of the original file
    -v (--verbose) - will show additional info about file sizes

Here is a full command that opens a file specified by `-t | --edit file`, using the key specified in `-k | --keyfile`, in the editor defined by the `$EDITOR` environment variable (or if not set – defaults to `/bin/vi`)".

To edit an encrypted file in `$EDITOR`, while asking to paste the key (`-i | --interactive`), while creating a backup file (`-b | --backup`):

    sym -tibf data.enc
    # => Private Key: ••••••••••••••••••••••••••••••••••••••••••••
    #
    # => Diff:
    # 3c3
    # # (c) 2015 Konstantin Gredeskoul.  All rights reserved.
    # ---
    # # (c) 2016 Konstantin Gredeskoul.  All rights reserved.


#### Using KeyChain Access on Mac OS-X

KeyChain storage is a huge time saver. It allows you to securely store the key the keychain, meaning the key can not be easily extracted by an attacker without a login to your account. Just having access to the disk is not enough.

Apple had released a `security` command line tool, which this library uses to securely store a key/value pair of the key name and the actual private key in your OS-X KeyChain. The advantages of this method are numerous:

 * The private key won't be lying around your file system unencrypted, so if your Mac is ever stolen, you don't need to worry about the keys running wild.
 * If you sync your keychain with the iCloud you will have access to it on other machines

As mentioned previously, to add the key to the KeyChain on the Mac, use `-x <key-name>` flag with `-g` flag when generating a key. The `key name` is what you call this particular key, based on how you plan to use it. For example, you may call it `staging`, etc.

The following command generates the private key and immediately stores it in the KeyChain access under the name provided:

    sym -gx staging   # the key is passwordless
    sym -gpcx staging # this key is password protected, with the password cached

Next, whenever you need to *use* this key, you can specify the key with `-k staging`. 

Finally, you can delete a key from KeyChain access by running:

    keychain <name> delete

Below we describe the purpose of the executable `keychain` shipped with sym.

#### KeyChain Key Management

`keychain` is an additional executable installed with the gem, which can be used to read (find), update (add), and delete keychain entries used by `sym`. 

It's help message is self-explanatory:

    Usage: keychain <name> [ add <contents> | find | delete ]

#### Moving a Key to the Keychain

You can easily move an existing key from a file or a string to a keychain by combining -k or -k to read the key, with -x to write it.

    sym -k $keysource -x mykey

#### Adding Password to Existing Key

You can add a password to a key by combining one of the key description flags (-k, -i) and then also -p.  Use `-q` to hide new key from the STDOUT, and `c` to cache the password.

    sym -k $mykey -pqcx moo
    
The above example will take an unencrypted key passed in `$mykey`, ask for a password and save password protected key into the keychain with name "moo."

#### Password Caching

Nobody likes to re-type passwords over and over again, and for this reason *Sym* supports password caching via a locally running `memcached` instance (using the default port 11211, if available).

*Multiple Providers*

Cache is written using the Provider design pattern (a.k.a. plugin architecture), and so it's easy to add a new Cache Provider that uses a custom backend. The supplied production-ready provider only works with a `memcached` daemon running (ideally) locally.


For customization of memcached location, we refer you to the `Configuration` class for an example of how to configure MemCached provider — shown below in the Ruby API section.

In order to control password caching, the following flags are available:

 * `-c` turns on caching
 * `-u seconds` sets the expiration for cached passwords
 * `-r memcached` controls which of the providers is used. Without this flag, *sym* auto-detects caching provider by first checking for `memcached`

#### Saving Common Flags in an Environment Variable


CHANGELOG.md
# Changelog

## [v3.0.2](https://github.com/kigster/sym/tree/v3.0.2) (2022-09-23)

[Full Changelog](https://github.com/kigster/sym/compare/v3.0.1...v3.0.2)

**Merged pull requests:**

- Adding github actions [\#33](https://github.com/kigster/sym/pull/33) ([kigster](https://github.com/kigster))

## [v3.0.1](https://github.com/kigster/sym/tree/v3.0.1) (2021-02-12)

[Full Changelog](https://github.com/kigster/sym/compare/v3.0.0...v3.0.1)

**Merged pull requests:**

- Fixing args expansion for Ruby 3.0.0 [\#32](https://github.com/kigster/sym/pull/32) ([kigster](https://github.com/kigster))

## [v3.0.0](https://github.com/kigster/sym/tree/v3.0.0) (2020-08-15)

[Full Changelog](https://github.com/kigster/sym/compare/v2.10.1...v3.0.0)

## [v2.10.1](https://github.com/kigster/sym/tree/v2.10.1) (2020-08-15)

[Full Changelog](https://github.com/kigster/sym/compare/list...v2.10.1)

## [list](https://github.com/kigster/sym/tree/list) (2020-08-15)

[Full Changelog](https://github.com/kigster/sym/compare/v2.10.0...list)

**Closed issues:**

- This gem is failing on environment without HOME \(like AWS Lambda\) [\#25](https://github.com/kigster/sym/issues/25)

**Merged pull requests:**

- Refactor and changes to support custom user\_home. [\#30](https://github.com/kigster/sym/pull/30) ([kigster](https://github.com/kigster))

## [v2.10.0](https://github.com/kigster/sym/tree/v2.10.0) (2020-08-14)

[Full Changelog](https://github.com/kigster/sym/compare/v2.8.5...v2.10.0)

**Fixed bugs:**

- First pass on Version 2.9.0 — dropping the coin gem [\#20](https://github.com/kigster/sym/pull/20) ([kigster](https://github.com/kigster))

**Closed issues:**

- Switch to Codecov for test coverage [\#29](https://github.com/kigster/sym/issues/29)
- Ruby 2.7 warnings [\#24](https://github.com/kigster/sym/issues/24)
- Error and process hang on Windows [\#17](https://github.com/kigster/sym/issues/17)

**Merged pull requests:**

- Switch to Codecov for test coverage; version bump 2.10.0 [\#28](https://github.com/kigster/sym/pull/28) ([kigster](https://github.com/kigster))
- Fix remaining 2.7 warnings; memcached & keychain [\#27](https://github.com/kigster/sym/pull/27) ([kigster](https://github.com/kigster))
- Adding Rubocop, and fixing minor issues. [\#21](https://github.com/kigster/sym/pull/21) ([kigster](https://github.com/kigster))
- CI: Update matrix [\#19](https://github.com/kigster/sym/pull/19) ([olleolleolle](https://github.com/olleolleolle))

## [v2.8.5](https://github.com/kigster/sym/tree/v2.8.5) (2018-10-13)

[Full Changelog](https://github.com/kigster/sym/compare/v2.8.4...v2.8.5)

**Closed issues:**

- \[feature\] Allow sym to use a password protected key without interactive input [\#12](https://github.com/kigster/sym/issues/12)
- \[feature\] support yaml mapping of file pattern to key names [\#8](https://github.com/kigster/sym/issues/8)

**Merged pull requests:**

- Fix error when ENV\[EDITOR\] is nil [\#18](https://github.com/kigster/sym/pull/18) ([JustinNazari](https://github.com/JustinNazari))
- README: Use SVG badge \[ci skip\] [\#16](https://github.com/kigster/sym/pull/16) ([olleolleolle](https://github.com/olleolleolle))

## [v2.8.4](https://github.com/kigster/sym/tree/v2.8.4) (2018-04-13)

[Full Changelog](https://github.com/kigster/sym/compare/v2.8.2...v2.8.4)

**Merged pull requests:**

- Faster completion by checking \_filedir [\#15](https://github.com/kigster/sym/pull/15) ([kigster](https://github.com/kigster))

## [v2.8.2](https://github.com/kigster/sym/tree/v2.8.2) (2018-01-10)

[Full Changelog](https://github.com/kigster/sym/compare/v2.8.1...v2.8.2)

## [v2.8.1](https://github.com/kigster/sym/tree/v2.8.1) (2018-01-07)

[Full Changelog](https://github.com/kigster/sym/compare/v2.8.0...v2.8.1)

**Merged pull requests:**

- Changing -t to expect a file argument; Version 2.7.0 [\#13](https://github.com/kigster/sym/pull/13) ([kigster](https://github.com/kigster))

## [v2.8.0](https://github.com/kigster/sym/tree/v2.8.0) (2018-01-06)

[Full Changelog](https://github.com/kigster/sym/compare/v2.7.0...v2.8.0)

## [v2.7.0](https://github.com/kigster/sym/tree/v2.7.0) (2017-06-23)

[Full Changelog](https://github.com/kigster/sym/compare/v2.6.3...v2.7.0)

## [v2.6.3](https://github.com/kigster/sym/tree/v2.6.3) (2017-03-13)

[Full Changelog](https://github.com/kigster/sym/compare/st...v2.6.3)

## [st](https://github.com/kigster/sym/tree/st) (2017-03-13)

[Full Changelog](https://github.com/kigster/sym/compare/v2.6.2...st)

**Closed issues:**

- \[feature\] Speed up RSpec tests by running Aruba tests in process, and skipping DRb [\#11](https://github.com/kigster/sym/issues/11)
- \[feature\] need an easy way to decrypt and load encrypted settings file into memory [\#9](https://github.com/kigster/sym/issues/9)

**Merged pull requests:**

- Switch to faster in process ARUBA [\#10](https://github.com/kigster/sym/pull/10) ([kigster](https://github.com/kigster))

## [v2.6.2](https://github.com/kigster/sym/tree/v2.6.2) (2017-03-12)

[Full Changelog](https://github.com/kigster/sym/compare/v2.6.1...v2.6.2)

## [v2.6.1](https://github.com/kigster/sym/tree/v2.6.1) (2017-03-12)

[Full Changelog](https://github.com/kigster/sym/compare/v2.6.0...v2.6.1)

## [v2.6.0](https://github.com/kigster/sym/tree/v2.6.0) (2017-03-12)

[Full Changelog](https://github.com/kigster/sym/compare/v2.5.3...v2.6.0)

## [v2.5.3](https://github.com/kigster/sym/tree/v2.5.3) (2017-03-11)

[Full Changelog](https://github.com/kigster/sym/compare/v2.5.2...v2.5.3)

## [v2.5.2](https://github.com/kigster/sym/tree/v2.5.2) (2017-03-11)

[Full Changelog](https://github.com/kigster/sym/compare/v2.5.1...v2.5.2)

## [v2.5.1](https://github.com/kigster/sym/tree/v2.5.1) (2017-03-07)

[Full Changelog](https://github.com/kigster/sym/compare/v2.5.0...v2.5.1)

## [v2.5.0](https://github.com/kigster/sym/tree/v2.5.0) (2017-03-05)

[Full Changelog](https://github.com/kigster/sym/compare/v2.4.3...v2.5.0)

## [v2.4.3](https://github.com/kigster/sym/tree/v2.4.3) (2017-03-01)


lib/sym.rb
require 'colored2'
require 'zlib'
require 'logger'

require 'sym/configuration'
require 'sym/constants'
require 'sym/version'
require 'sym/errors'

Sym::Configuration.configure do |config|
  config.password_cipher          = 'AES-128-CBC'
  config.data_cipher              = 'AES-256-CBC'
  config.private_key_cipher       = config.data_cipher
  config.compression_enabled      = true
  config.compression_level        = Zlib::BEST_COMPRESSION
  config.encrypted_file_extension = 'enc'
  config.default_key_file         = Sym::Constants.sym_key_file

  config.password_cache_timeout          = 300

  # When nil is selected, providers are auto-detected.
  config.password_cache_default_provider = nil
  config.password_cache_arguments        = {
    memcached: {
      args: %w(127.0.0.1:11211),
      opts: { namespace:  'sym',
              compress:   true,
              expires_in: config.password_cache_timeout}

    }
  }
end

require 'sym/extensions/stdlib'
require 'sym/extensions/class_methods'
require 'sym/extensions/instance_methods'
#
# == Using Sym Library
#
# This library is a "wrapper" that allows you to take advantage of the
# symmetric encryption functionality provided by the {OpenSSL} gem (and the
# underlying C library). In order to use the library in your ruby classes, you
# should _include_ the module {Sym}.
#
# The including class is decorated with four instance methods from the
# module {Sym::Extensions::InstanceMethods} and two class methods from
# {Sym::Extensions::ClassMethods} – for specifics, please refer there.
#
# The two main instance methods are +#encr+ and +#decr+, which as the name
# implies, perform two-way symmetric encryption and decryption of any Ruby object
# that can be +marshaled+.
#
# Two additional instance methods +#encr_password+ and +#decr_password+ turn on
# password-based encryption, which actually uses a password to construct a 128-bit
# long private key, and then uses that in the encryption of the data.
# You could use them to encrypt data with a password instead of a randomly
# generated private key.
#
# The library comes with a rich CLI interface, which is mostly encapsulated under the
# +Sym::App+ namespace.
#
# The +sym+ executable that is the "app" in this case, and is a _user_ of the
# API methods +#encr+ and +#decr+.
#
# Create a new key with +#create_private_key+ class method, which returns a new
# key every time it's called, or with +#private_key+ class method, which either
# assigns, or creates and caches the private key at a class level.
#
# == Example
#
#     require 'sym'
#
#     class TestClass
#       include Sym
#       # read the key from environmant variable and assign to this class.
#       private_key ENV['PRIVATE_KEY']
#
#       def sensitive_value=(value)
#         @sensitive_value = encr(value, self.class.private_key)
#       end
#
#       def sensitive_value
#         decr(@sensitive_value, self.class.private_key)
#       end
#     end
#
# == Private Key
#
# They private key can be generated by +TestClass.create_private_key+
# which returns but does not store a new random 256-bit key.
#
# The key can be assigned and saved, or auto-generated and saved using the
# +#private_key+ method on the class that includes the +Sym+ module.
#
# Each class including the +Sym+ module would get their own +#private_key#
# class-instance variable accessor, and a possible value.
#

module Sym
  def self.included(klass)
    klass.instance_eval do
      include ::Sym::Extensions::InstanceMethods
      extend ::Sym::Extensions::ClassMethods
      class << self
        def private_key(value = nil)
          if value
            @private_key= value
          elsif @private_key
            @private_key
          else
            @private_key= self.create_private_key
          end
          @private_key
        end
      end
    end
  end

  class << self
    def config
      Sym::Configuration.config
    end

    def default_key_file
      config.default_key_file
    end

    def default_key
      File.read(default_key_file) rescue nil
    end

    def default_key?
      File.exist?(default_key_file)
    end

  end
end

